libraries/AP_NavEKF3/AP_NavEKF3.cpp

```diff
@@ -1308,10 +1318,9 @@ bool NavEKF3::setOriginLLH(const Location &loc)
     if (!core) {
         return false;
     }
-    if (_fusionModeGPS != 3) {
-        // we don't allow setting of the EKF origin unless we are
-        // flying in non-GPS mode. This is to prevent accidental set
-        // of EKF origin with invalid position or height
+    if (_sources.getPosXYSource() == AP_NavEKF_Source::SourceXY::GPS) {
```

Comment:
nice catch.  .. but if this is an issue then it is an existing issue because the code this replaces also does not check that the ALT_SOURCE != 2 (GPS).

I think we should not add a check for whether GPS is being used for altitude because if we did, we would indirectly make it impossible for users to use GPS for alt + optical flow for horizontal.  This is a very odd combination but I don't think we should disallow this.  The danger is that users could end up with the ekf origin very far from the home altitude and if this is greater than 100km they could lose some precision.  I've raised an issue to add a pre-arm check.  https://github.com/ArduPilot/ardupilot/issues/15800